src/ripple/app/misc/impl/AMMHelpers.cpp

```diff
@@ -31,61 +31,94 @@ ammLPTokens(
     STAmount const& asset1Deposit,
-    STAmount const& lpTokensBalance,
+    STAmount const& lptAMMBalance,
     std::uint16_t tfee)
 {
-    return toSTAmount(
-        lpTokensBalance.issue(),
-        lpTokensBalance *
-            (root2(1 + (asset1Deposit * feeMultHalf(tfee)) / asset1Balance) -
-             1));
+    auto const f1 = feeMult(tfee);
+    auto const f2 = feeMultHalf(tfee) / f1;
+    Number const r = asset1Deposit / asset1Balance;
+    auto const c = root2(f2 * f2 + r / f1) - f2;
+    auto const t = lptAMMBalance * (r - c) / (1 + c);
+    return toSTAmount(lptAMMBalance.issue(), t);
 }
 
+/* Equation 4 solves equation 3 for b:
+ * Let f1 = 1 - tfee, f2 = (1 - tfee/2)/f1, t1 = t/T, t2 = 1 + t1, R = b/B
+ * then
+ * t1 = [R - sqrt(f2**2 + R/f1) + f2] / [1 + sqrt(f2**2 + R/f1] - f2] =>
+ * sqrt(f2**2 + R/f1)*(t1 + 1) = R + f2 + t1*f2 - t1 =>
+ * sqrt(f2**2 + R/f1)*t2 = R + t2*f2 - t1 =>
+ * sqrt(f2**2 + R/f1) = R/t2 + f2 - t1/t2, let d = f2 - t1/t2 =>
+ * sqrt(f2**2 + R/f1) = R/t2 + d => f2**2 + R/f1 = (R/t2)**2 +2*d*R/t2 + d**2 =>
+ * (R/t2)**2 + R*(2*d/t2 - 1/f1) + d**2 - f2**2 = 0
+ */
 STAmount
 ammAssetIn(
     STAmount const& asset1Balance,
-    STAmount const& lpTokensBalance,
     STAmount const& lptAMMBalance,
+    STAmount const& lpTokens,
     std::uint16_t tfee)
 {
+    auto const f1 = feeMult(tfee);
+    auto const f2 = feeMultHalf(tfee) / f1;
+    auto const t1 = lpTokens / lptAMMBalance;
+    auto const t2 = 1 + t1;
+    auto const d = f2 - t1 / t2;
+    auto const a = 1 / (t2 * t2);
+    auto const b = 2 * d / t2 - 1 / f1;
+    auto const c = d * d - f2 * f2;
     return toSTAmount(
-        asset1Balance.issue(),
-        ((square(lpTokensBalance / lptAMMBalance + 1) - 1) /
-         feeMultHalf(tfee)) *
-            asset1Balance);
+        asset1Balance.issue(), asset1Balance * solveQuadraticEq(a, b, c));
 }
 
+/* Equation 7:
+ * t = T * (c - sqrt(c**2 - 4*R))/2
+ * where R = b/B, c = R*fee + 2 - fee
+ */
 STAmount
 lpTokensOut(
     STAmount const& asset1Balance,
     STAmount const& asset1Withdraw,
-    STAmount const& lpTokensBalance,
+    STAmount const& lptAMMBalance,
     std::uint16_t tfee)
 {
-    if (auto const a =
-            Number(1) - asset1Withdraw / (asset1Balance * feeMultHalf(tfee));
-        a <= 0 || a >= 1)
-        return STAmount{};
-    else
-        return toSTAmount(
-            lpTokensBalance.issue(), lpTokensBalance * (1 - root2(a)));
+    Number const fr = asset1Withdraw / asset1Balance;
+    auto const f1 = getFee(tfee);
+    auto const c = fr * f1 + 2 - f1;
+    auto const t = lptAMMBalance * (c - root2(c * c - 4 * fr)) / 2;
+    return toSTAmount(lptAMMBalance.issue(), t);
 }
 
+/* Equation 8 solves equation 7 for b:
+ * c - 2*t/T = sqrt(c**2 - 4*R) =>
+ * c**2 - 4*c*t/T + 4*t**2/T**2 = c**2 - 4*R =>
+ * -4*c*t/T + 4*t**2/T**2 = -4*R => -c*t/T + t**2/T**2 = -R
+ * substitute c = R*f +2 - f =>
+ * -t**R*f + 2 - f) + t**2/T**2 = -R, let t1 = t/T =>
```

Comment:
Two things:
1. I think this would be easier to read with each equation on its own line.
2. What's happening here? `-t**R*f + 2 - f)`. I would expect `-t/T*(R*f + 2 -f)`